File: noisecluster-win/noisecluster/win/transport/aeron/AeronTarget.cs
/**
  * Copyright 2017 https://github.com/sndnv
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  * http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */

using System;
using System.Threading;
using Adaptive.Aeron;
using Adaptive.Agrona.Concurrent;
using log4net;

namespace noisecluster.win.transport.aeron
{
    public class AeronTarget : IDisposable
    {
        private readonly ILog _log = LogManager.GetLogger(typeof(AeronTarget));
        private int _isRunning; //0 = false; 1 = true
        private readonly int _stream;
        private readonly string _channel;
        private readonly Subscription _subscription;
        private readonly IIdleStrategy _idleStrategy;
        private readonly int _fragmentLimit;

        public bool IsActive
        {
            get { return _isRunning == 1; }
        }

        public delegate void DataHandler(byte[] data, int length);

        public AeronTarget(
            Aeron aeron,
            int stream,
            string channel,
            IIdleStrategy idleStrategy,
            int fragmentLimit
        )
        {
            _stream = stream;
            _channel = channel;
            _subscription = aeron.AddSubscription(_channel, _stream);
            _idleStrategy = idleStrategy;
            _fragmentLimit = fragmentLimit;
        }

        //docs - warn about blocking
        public void Start(DataHandler dataHandler)
        {
            if (Interlocked.CompareExchange(ref _isRunning, 1, 0) == 0)
            {
                _log.InfoFormat("Starting transport for channel [{0}] and stream [{1}]", _channel, _stream);
                var fragmentAssembler = new FragmentAssembler(
                    (buffer, offset, length, _) =>
                    {
                        var data = new byte[length];
                        buffer.GetBytes(offset, data);
                        dataHandler(data, length);
                    }
                );

                //will block until stopped
                while (_isRunning == 1)
                {
                    var fragmentsRead = _subscription.Poll(fragmentAssembler.OnFragment, _fragmentLimit);
                    _idleStrategy.Idle(fragmentsRead);
                }

                _log.InfoFormat("Stopped transport for channel [{0}] and stream [{1}]", _channel, _stream);
            }
            else
            {
                var message = string.Format(
                    "Cannot start transport for channel [{0}] and stream [{1}]; transport is already active",
                    _channel,
                    _stream
                );
                _log.Warn(message);
                throw new InvalidOperationException(message);
            }
        }

        public void Stop()
        {
            if (Interlocked.CompareExchange(ref _isRunning, 0, 1) == 1)
            {
                _log.InfoFormat("Stopping transport for channel [{0}] and stream [{1}]", _channel, _stream);
            }
            else
            {
                var message = string.Format(
                    "Cannot stop transport for channel [{0}] and stream [{1}]; transport is not active",
                    _channel,
                    _stream
                );
                _log.Warn(message);
                throw new InvalidOperationException(message);
            }
        }

        public void Close()
        {
            if (_isRunning == 0)
            {
                _log.InfoFormat("Closing transport for channel [{0}] and stream [{1}]", _channel, _stream);
                _subscription.Dispose();
                _log.InfoFormat("Closed transport for channel [{0}] and stream [{1}]", _channel, _stream);
            }
            else
            {
                var message = string.Format(
                    "Cannot close transport for channel [{0}] and stream [{1}]; transport is still active",
                    _channel,
                    _stream
                );
                _log.Warn(message);
                throw new InvalidOperationException(message);
            }
        }

        public void Dispose()
        {
            _subscription.Dispose();
        }
    }
}

File: noisecluster-jvm/src/main/scala/noisecluster/jvm/transport/aeron/AeronTarget.scala
/**
  * Copyright 2017 https://github.com/sndnv
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  * http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
package noisecluster.jvm.transport.aeron

import java.util.concurrent.atomic.AtomicBoolean

import akka.actor.ActorSystem
import akka.event.Logging
import io.aeron._
import io.aeron.logbuffer._
import noisecluster.jvm.transport.Target
import org.agrona.DirectBuffer
import org.agrona.concurrent._

import scala.concurrent.blocking

class AeronTarget(
  private val stream: Int,
  private val channel: String,
  private val idleStrategy: IdleStrategy,
  private val fragmentLimit: Int
)(implicit loggingActorSystem: ActorSystem, aeron: Aeron) extends Target {
  private val isRunning = new AtomicBoolean(false)
  private val log = Logging.getLogger(loggingActorSystem, this)
  private val subscription = aeron.addSubscription(channel, stream)

  def isActive: Boolean = isRunning.get

  //docs - warn about blocking
  override def start(dataHandler: (Array[Byte], Int) => Unit): Unit = {
    if (isRunning.compareAndSet(false, true)) {
      log.info("Starting transport for channel [{}] and stream [{}]", channel, stream)

      val fragmentAssembler = new FragmentAssembler(
        (buffer: DirectBuffer, offset: Int, length: Int, _: Header) => {
          val data = new Array[Byte](length)
          buffer.getBytes(offset, data)
          dataHandler(data, length)
        }
      )

      blocking {
        while (isRunning.get) {
          val fragmentsRead = subscription.poll(fragmentAssembler, fragmentLimit)
          idleStrategy.idle(fragmentsRead)
        }
      }

      log.info("Stopped transport for channel [{}] and stream [{}]", channel, stream)

    } else {
      val message = s"Cannot start transport for channel [$channel] and stream [$stream]; transport is already active"
      log.warning(message)
      throw new IllegalStateException(message)
    }
  }

  override def stop(): Unit = {
    if (isRunning.compareAndSet(true, false)) {
      log.info("Stopping transport for channel [{}] and stream [{}]", channel, stream)
    } else {
      val message = s"Cannot stop transport for channel [$channel] and stream [$stream]; transport is not active"
      log.warning(message)
      throw new IllegalStateException(message)
    }
  }

  override def close(): Unit = {
    if (!isRunning.get) {
      log.info("Closing transport for channel [{}] and stream [{}]", channel, stream)
      subscription.close()
      log.info("Closed transport for channel [{}] and stream [{}]", channel, stream)
    } else {
      val message = s"Cannot close transport for channel [$channel] and stream [$stream]; transport is still active"
      log.warning(message)
      throw new IllegalStateException(message)
    }
  }
}

object AeronTarget {
  def apply(
    stream: Int,
    address: String,
    port: Int,
    idleStrategy: IdleStrategy,
    fragmentLimit: Int
  )(implicit loggingActorSystem: ActorSystem, aeron: Aeron): AeronTarget =
    new AeronTarget(
      stream,
      s"aeron:udp?endpoint=$address:$port",
      idleStrategy,
      fragmentLimit
    )

  def apply(
    stream: Int,
    address: String,
    port: Int,
    interface: String,
    idleStrategy: IdleStrategy,
    fragmentLimit: Int
  )(implicit loggingActorSystem: ActorSystem, aeron: Aeron): AeronTarget =
    new AeronTarget(
      stream,
      s"aeron:udp?endpoint=$address:$port|interface=$interface",
      idleStrategy,
      fragmentLimit
    )

  def apply(
    stream: Int,
    channel: String,
    idleStrategy: IdleStrategy,
    fragmentLimit: Int
  )(implicit loggingActorSystem: ActorSystem, aeron: Aeron): AeronTarget =
    new AeronTarget(
      stream,
      channel,
      idleStrategy,
      fragmentLimit
    )

  def apply(
    stream: Int,
    address: String,
    port: Int
  )(implicit loggingActorSystem: ActorSystem, aeron: Aeron): AeronTarget =
    AeronTarget(
      stream,
      address,
      port,
      Defaults.IdleStrategy,
      Defaults.FragmentLimit
    )
}
